pkg/distributor/distributor_ingest_storage_test.go

```diff
@@ -0,0 +1,319 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package distributor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/grafana/dskit/grpcutil"
+	"github.com/grafana/dskit/mtime"
+	"github.com/grafana/dskit/user"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/pkg/kmsg"
+	"google.golang.org/grpc/codes"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util/extract"
+	"github.com/grafana/mimir/pkg/util/testkafka"
+)
+
+// kafkaTopic is the Kafka topic used for ingest storage tests.
+const kafkaTopic = "test"
+
+func TestDistributor_Push_ShouldSupportIngestStorage(t *testing.T) {
+	ctx := user.InjectOrgID(context.Background(), "user")
+
+	// Mock distributor current time (used to get stable metrics assertion).
+	now := time.Now()
+	mtime.NowForce(now)
+	t.Cleanup(func() {
+		mtime.NowReset()
+	})
+
+	// To keep assertions simple, all tests send the same request.
+	createRequest := func() *mimirpb.WriteRequest {
+		return &mimirpb.WriteRequest{
+			Timeseries: []mimirpb.PreallocTimeseries{
+				makeTimeseries([]string{model.MetricNameLabel, "series_one"}, makeSamples(now.UnixMilli(), 1), makeExemplars([]string{"trace_id", "xxx"}, now.UnixMilli(), 1)),
+				makeTimeseries([]string{model.MetricNameLabel, "series_two"}, makeSamples(now.UnixMilli(), 2), nil),
+				makeTimeseries([]string{model.MetricNameLabel, "series_three"}, makeSamples(now.UnixMilli(), 3), nil),
+				makeTimeseries([]string{model.MetricNameLabel, "series_four"}, makeSamples(now.UnixMilli(), 4), nil),
+				makeTimeseries([]string{model.MetricNameLabel, "series_five"}, makeSamples(now.UnixMilli(), 5), nil),
+			},
+			Metadata: []*mimirpb.MetricMetadata{
+				{MetricFamilyName: "series_one", Type: mimirpb.COUNTER, Help: "Series one description"},
+				{MetricFamilyName: "series_two", Type: mimirpb.COUNTER, Help: "Series two description"},
+			},
+		}
+	}
```

Comment:
can you use `makeWriteRequest` instead of implementing it here too?

Reply:
`makeWriteRequest()` is one of those functions I don't like very much because when you look at a test it's not clear what's going to be written. I prefer to keep it as is.